smart-contracts/wasm-chain-integration/src/v1/mod.rs

```diff
@@ -309,6 +325,20 @@ mod host {
                     Err(e) => bail!("Illegal call, cannot parse arguments: {:?}", e),
                 }
             }
+            UPGRADE_TAG => {
+                ensure!(
+                    length == MODULE_REF_SIZE,
+                    "Upgrade must have exactly 32 bytes of payload, but was {}",
+                    length
+                );
+                ensure!(start + length <= memory.len(), "Illegal memory access.");
+                let mut module_ref_bytes = [0u8; MODULE_REF_SIZE];
+                module_ref_bytes.copy_from_slice(&memory[start..start + MODULE_REF_SIZE]);
+                let module_ref = ModuleRef(module_ref_bytes);
+                Ok(Some(Interrupt::Upgrade {
+                    module_ref,
+                }))
+            }
```

Comment:
The `upgrade` should not be exposed as an operation in this way. The reason for this is that this way there is no practical way of checking "is it upgradable". We need to add a new host function.

However internally we can handle it as an operation like you did (internally being in the scheduler and FFI).